src/db/gc/mod.rs

```diff
@@ -0,0 +1,309 @@
+    /// * `get_heaviest_tipset` - A function that facilitates heaviest tipset retrieval.
+    /// * `depth` - The number of state-roots to retain.
+    /// * `block_time` - An average block production time.
+    pub fn new(
+        db: Arc<DB>,
+        get_heaviest_tipset: Box<dyn Fn() -> Arc<Tipset> + Send>,
+        depth: ChainEpochDelta,
+        block_time: Duration,
+    ) -> Self {
+        Self {
+            db,
+            get_heaviest_tipset,
+            depth,
+            marked: HashSet::new(),
+            epoch_marked: 0,
+            block_time,
+        }
+    }
+    // Populate the initial set with all the available database keys.
+    fn populate(&mut self) -> anyhow::Result<()> {
+        self.marked = self.db.get_keys()?;
+        Ok(())
+    }
+
+    // Filter out the initial set, leaving only the entries that need to be removed.
+    // NOTE: One concern here is that this is going to consume a lot of CPU.
+    async fn filter(&mut self, tipset: Arc<Tipset>, depth: ChainEpochDelta) -> anyhow::Result<()> {
+        // NOTE: We want to keep all the block headers from genesis to heaviest tipset epoch.
+        let mut stream = stream_graph(
+            self.db.clone(),
+            (*tipset).clone().chain(self.db.clone()),
+            depth,
+        );
+
+        while let Some(block) = stream.next().await {
+            let block = block?;
+            self.marked.remove(&truncated_hash(block.cid.hash()));
+        }
+
+        anyhow::Ok(())
+    }
+
+    // Remove marked keys from the database.
+    fn sweep(&mut self) -> anyhow::Result<()> {
+        let marked = mem::take(&mut self.marked);
+        self.db.remove_keys(marked)
+    }
+
+    /// Starts the Garbage Collection loop.
+    ///
+    /// # Arguments
+    ///
+    /// * `depth` - Specifies how far back the full history should be maintained. Cannot be less
+    /// than chain finality.
+    /// * `interval` - GC Interval to avoid constantly consuming node's resources.
+    ///
+    /// NOTE: This currently does not take into account the fact that we might be starting the node
+    /// using CAR-backed storage with a snapshot, for implementation simplicity.
+    pub async fn gc_loop(&mut self, interval: Duration) -> anyhow::Result<()> {
+        loop {
+            self.gc_workflow(interval).await?
+        }
+    }
+
+    // This function yields to the main GC loop if the conditions are not met for execution of the
+    // next step.
+    async fn gc_workflow(&mut self, interval: Duration) -> anyhow::Result<()> {
+        let depth = self.depth;
+        let tipset = (self.get_heaviest_tipset)();
+        let current_epoch = tipset.epoch();
+        // Don't run the GC if there aren't enough state-roots yet. Sleep and yield to the main loop
+        // in order to refresh the heaviest tipset value.
+        if depth > current_epoch {
+            time::sleep(self.block_time * (depth - current_epoch) as u32).await;
```

Comment:
If `depth > current_epoch`, you can just sleep for `interval` amount of time. No need to be any more precise than that.